### PR TITLE
Instagram

### LINKED ISSUE #7: Support Instagram
By request of @codepunkt: add support for Instagram.

https://www.instagram.com/developer/embedding/#oembed

### PR DESCRIPTION
**Pull request by FabioRosado**
**What**: Added Instagram support 

**Why**:  Fixes #7

**How**: Added instagram.js and updated tests

**Checklist**:

- [x] Documentation
- [x] Tests
- [x] Ready to be merged
      

Hello,
this is my first time contributing to a React/js project and I've been a bit confused by the tests, one test is still failing but I don't understand what is wrong - I'm going to assume it might be something related to spaces but i'm not sure.

This seems like an awesome project and I want to use this plugin on my gatsby sites so decided to help 😄 

On this PR I had to mix the html code that instagram returns with the other params since they don't include the image on the embedded widget which is a shame. They use a lot of styles to style the widget, I'm not sure if you are happy with adding them here or not, but I decided to keep the code as close to the original html one as possible.

### CODE REVIEW

**Comment by MichaelDeBoey** [comment]
Hi @FabioRosado!

Thanks for taking your time to implement the Instagram transformer! 🙂

I'm currently on holiday until mid January, so I won't be able to review this one quickly.
Will try to look at it asap. 🙂

**Comment by FabioRosado** [comment]
No worries I'm not in a hurry haha enjoy your holidays!

**Comment by MichaelDeBoey** [comment]
@FabioRosado Can you give a reproduction repo or CodeSandbox link where I can see the difference between the original html the `oembed` api is returning and the one you're returning please?

**Comment by FabioRosado** [comment]
Actually I just realised that if we allow the script tag then instagram will load the picture just fine. Would you be happy to just use the instagram html or should we keep the script out? I was looking at how the Twitter transformer was implemented and used the `omit_script=true` because of that 🤔

**Comment by FabioRosado** [comment]
I've updated the PR to use the html code that instagram returns and should be okay.

You can see the demo of the html returned by the instagram omebed here: [codesandbox](https://codesandbox.io/s/gatsby-remark-embedder-instagram-2uc2d)

I'm wondering if the script pointing to `src="\\www.instagram.com/embeded.js"` will cause some issues and if we should remove the backslashes 🤔 

I've also tried to test this plugin on a new gatsby site, but couldn't make it work. Is it possible to manual test the plugin like this? 🤔 

This was my config:
```{
      resolve: `gatsby-transformer-remark`,
      options: {
        plugins: [
          {
            resolve: require.resolve(`<path to remark-embedded folder>`)
          },
```

Thank you for the time and sorry about all the questions 😄

**Comment by osiux** [comment]
Hi @FabioRosado, nice work! I was wondering if the instagram ebedder should use the same approach the [Twitter embedder uses](https://github.com/MichaelDeBoey/gatsby-remark-embedder#twitter), which is using an external plugin to load the embed js code. The reason for this is the same as for the twitter one, we don't want to include the snippet several times if several instagram posts are included in the same page.

Maybe use [gatsby-plugin-instagram-embed](https://github.com/jlengstorf/gatsby-plugin-instagram-embed)? This would required to add the `omitscript` param to the oembed url and maybe change some classes, or create a new plugin that mimics the same behavior as the one for twitter and just adds the code once to the site.

What do you think?

**Comment by FabioRosado** [comment]
Hello @osiux thank you for the comment. I was actually thinking about this when I started working on the PR and did try to copy instagram html to show the picture and not use the script.

I'm not sure if people would want to show more than one instagram picture in a single post, but if that is the case then I agree doing it like the twitter one uses.

I'd like to know your opinion on this matter Michael and let me know if you would like me to change the code to use a custom made one and follow the lines of the Twitter one 😄 👍

**Comment by MichaelDeBoey** [comment]
@FabioRosado I agree with @osiux that we shouldn't add the script (so using `omitscript=true`) and we should recommend @jlengstorf's [`gatsby-plugin-instagram-embed`](https://github.com/jlengstorf/gatsby-plugin-instagram-embed).

> The returned HTML snippet from the Instagram transformer will only be automatically recognized as an embedded post when Instagram's embed JavaScript is included on the page.  
> Since the Instagram transformer doesn't include this JavaScript (because we don't want to include it multiple times on a page when having multiple embeds), you have to include it yourself. The recommended way of including it is by using [`gatsby-plugin-instagram-embed`](https://github.com/jlengstorf/gatsby-plugin-instagram-embed).

I'll review your PR when I have some more time this week (or early next week) btw 🙂

**Comment by FabioRosado** [comment]
Sounds good to me I will start working on these changes now 😄 👍

**Comment by acagastya** [comment]
I was wondering if using the blockquote method was the best way to deal with this case.  I think I have found a much simpler way to solve this without the hassle of using oembed or fetching the data

@FabioRosado, would it be okay for me to send a PR implementing that?

**Comment by MichaelDeBoey** [comment]
Sure @acagastya if you think a totally different approach is a better one, be my guest to create a PR explaining why you think we should go that way instead. 🙂

**Comment by FabioRosado** [comment]
I was just following the directions of the other transformers since they use the blockquote method. What is you idea?

**Comment by MichaelDeBoey** [comment]
@FabioRosado I haven't looked at your PR yet, but if you followed the same directions as the other transformers, that's how I would want it to be.

But maybe @acagastya has a better idea of how we can make our transformers.
I'm always open for suggestions. 🙂

**Comment by acagastya** [comment]
@FabioRosado @MichaelDeBoey Looks like using fetch and the API was inevitable, but we can work on the blockquote.

Have a look at it here:

https://gist.github.com/acagastya/684991cbe14db62968dce189a8c93686

**Comment by FabioRosado** [comment]
I've had a look at that gist and looks like you are using the url param returned by the api right? Wouldn't the script tag being imported every time we embed a picture  even if it's in an iframe? 

I am just curious, why do you think using blockquote is an issue? This is what the instagram api call returns and so do the other transformers.

Would be happy to discuss this further and see what would be best for the project 😄 👍

**Comment by acagastya** [comment]
There are other transformations which don't use the blockquote structure: the only concern is the blockquote adds too much for the same thing which iframe accomplishes.  However, if script has to be avoided at all costs, well there is there is `/media?size=l` option to get the image.

File: src/transformers/Instagram.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import { URL } from 'url';
+import fetch from 'node-fetch';
+
+export const shouldTransform = url => {
+  const { host, pathname } = new URL(url);
+
+  return (
+    ['instagr.am', 'instagram.com', 'www.instagram.com'].includes(host) &&
```
Comment (MichaelDeBoey) [inline src/transformers/Instagram.js]: Add support for `www.instagr.am` too

**Comment by FabioRosado** [comment]
Thanks for the review I'll go and fix these things. The only reason why I decided to recreate the html is because if you forget to include the instagram script no image will be shown when you try to embed the image - Not sure if you seen the result on the codebox that I've linked on my previous comments

**Comment by MichaelDeBoey** [comment]
Hi @FabioRosado!

I wouldn't bother that.
We're explicitly explaining in the docs that you should use `gatsby-plugin-instagram-embed` if you're using the Instagram transformer (just like we do with the Twitter transformer)

**Comment by FabioRosado** [comment]
Apologies it took me longer than I wanted, I've added all the requested changes and the tests are passing 👍

**Comment by MichaelDeBoey** [comment]
@all-contributors Please add @FabioRosado for code and tests

**Pull request merged**